### PR TITLE
feat(config)!: use object parameter instead of array

### LINKED ISSUE #8: Use paddings config object instead of array
The addon uses an array to define padding options, similar to the [backgrounds addon](https://github.com/storybookjs/storybook/tree/master/addons/backgrounds). This makes it difficult to merge global configurations to story overrides, forcing the redefinition of global values in stories.

To avoid this issue, the addon should use a config object (map) instead, like the [viewport addon](https://github.com/storybookjs/storybook/tree/master/addons/viewport):

```js
addParameters({
  paddings: {
    paddings: {
      small: { name: 'Small', value: '16px' },
      medium: { name: 'Medium', value: '32px' },
      largin: { name: 'Large', value: '64px' },
    },
    defaultPadding: 'medium',
  },
});
```

### PR DESCRIPTION
**Pull request by joanrm20**
Hey @rbardini,

So I started to work on #8. Let me know if this is in the right path of what you were aiming to solve. :) 

What I did:
1. In order to be able to keep array and object config I'm normalizing just the entries/items so the rest of the internal logic will work in the same way.
2. I'm proposing the keyword `sizes` as part of the `paddings` global object, just so it looks a bit more readable from the config outside. Please let me know what do you think on this one? :) 
```
{
  paddings: {
    sizes: {
      small: { name: 'Small', value: '16px' },
      medium: { name: 'Medium', value: '32px' },
      largin: { name: 'Large', value: '64px' },
    },
    defaultPadding: 'medium',
  }
}
```

Pending:
1. I still would like to know how you'd like to approach the config from the examples and docs point of view? Should I create a different example? Set the example to object as default or simply just update the docs with the new option?

Alternative Option:
1. Instead of normalizing the entries we could diverge the logic and treat it internally as object or array for every condition. Although this might require some more code as it will need a validation for every condition checking whether it's the Array one or the Object and do finds or object "picks" from the provisioned configurations. I can provide code example of this one if needed. :)
2. In the same line of thought of 1, I found a small concern on that approach. Even if we use object notation sometime to find "x" value we'll still might need to turn the config into an array to find the previously picked item.

### CODE REVIEW

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: I think we should follow the [new Backgrounds addon API](https://github.com/storybookjs/storybook/blob/badece61154f4cd372404786186125a7369e362f/addons/backgrounds/src/containers/BackgroundSelector.tsx#L35-L39), with `values`, `default` and `disable` fields.

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: Must [check the value of `disable` field](https://github.com/storybookjs/storybook/blob/badece61154f4cd372404786186125a7369e362f/addons/backgrounds/src/containers/BackgroundSelector.tsx#L128-L129) as well when it's an object. This wasn't needed before because anything other than a non-empty array was considered disabled.

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: Once we update the field names, we can rename these `values` and `normalizeValues`.

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: We should display a deprecation warning in this case, to help users migrate to the new API. We can use Storybook's logger for this, see [example.](https://github.com/storybookjs/storybook/blob/badece61154f4cd372404786186125a7369e362f/addons/backgrounds/src/containers/BackgroundSelector.tsx#L122-L126)

**Review by rbardini — CHANGES_REQUESTED** [review]
Thanks, @joanrm20! I added a few comments, mostly to follow the conventions of the [new Backgrounds addon API](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#backgrounds-addon-has-a-new-api).

As for the docs, you can update all README and demo examples to use the object format. Migration tips will be added to the GitHub release once merged.

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: Done:

File: src/containers/PaddingSelector.tsx
Comment (joanrm20) [inline src/containers/PaddingSelector.tsx:61]: hey @rbardini, I removed this validation as it's not necessary given the outer validation of isEnabled which already takes into account the items.

File: src/helpers.tsx
Comment (joanrm20) [inline src/helpers.tsx]: hey @rbardini, I think the logic inside `getSelectedPadding` was getting a bit complex given the access of other helper functions.
What I did is:
1. Ensure it always received the same collection of elements it needs, to pick a padding.
2. The consumers `index` and `register`, will ensure to validate`(isEnable)` and pass the right data `(normalizeValues)`.

This avoids the need of calling helpers internally and in the consumers as well.

File: src/helpers.tsx
Comment (joanrm20) [inline src/helpers.tsx]: @rbardini Please, let me know your thoughts on the message 😬

File: src/containers/PaddingSelector.tsx
Comment (rbardini) [reply src/containers/PaddingSelector.tsx:61]: Makes sense 🙂

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: I think it's better to rename this `ValueMap`. However, I just noticed that the new Backgrounds addon API is not using a map, like the Viewports addon, but rather an object with an array and a default value. Maybe we should do the same?

It's more likely that consumers would change the default background of a story, instead of the values. This should also simplify the implementation and types, so `PaddingConfig` would look like [this](https://github.com/storybookjs/storybook/blob/aba588cdc779deb5dd9c41046b0c193b7292d659/addons/backgrounds/src/containers/BackgroundSelector.tsx#L35-L39), with `values` declared inline.

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: It's probably a good idea to rename `Option` to `ValueWithDefault`. However, if we change the API (see comments above), it could look like this instead:

```tsx
type Padding = { name: string; value: string };

type PaddingsParameter = {
  default?: string;
  disable?: boolean;
  values: Padding[];
}

type PaddingWithDefault = Padding & Pick<PaddingsParameter, 'default'>;
```

It should also help avoiding confusion between these `Value` and `Option` types.

File: src/containers/PaddingSelector.tsx
Comment (rbardini) [inline src/containers/PaddingSelector.tsx]: Wouldn't it be better to call `normalizeValues` first and pass the normalized values to `isEnabled`? This way we don't need to check if `options` is an array in both places.

File: src/helpers.tsx
Comment (rbardini) [reply src/helpers.tsx]: I would suggest something like:

> Using array parameter with Paddings addon is deprecated, and support will be dropped in a future major version. Please refer to the release notes for migration: https://github.com/rbardini/storybook-addon-paddings/releases/tag/v2.1.0

(assuming v2.1.0 will be the next release)

File: src/helpers.tsx
Comment (rbardini) [reply src/helpers.tsx]: Sounds good 👍

**Review by rbardini — CHANGES_REQUESTED** [review]
@joanrm20 I added a few more suggestions, especially regarding the API. Let me know what you think 🙂

Also, we need to update the README examples, but it's better to decide the final API first.

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: hey @rbardini, yeah when I was working over the actual implementation I noticed it too. The couple reasons I didn't went trough that implementation before:

1. I was still under the object sort of line of thought since previous discussions.
2. I liked a bit more the idea of having a defined key to use per entry as a filter value to set which is the default. Thinking maybe to avoid possible clashes or difficulties with verbose names. Still I don't know how often it may happen or if would actually be an use case. 
3. Keys used to override default are shorter for users. 

e.g: 
`{name: 'Extra Large Mobile', value: '42px'}`, then the consumer to override in other places will need the name. Which could be solve by:
1. Keeping the object notation
2. Switching to array and add a key to the entry `{key: 'xl-mobile', name: 'Extra Large Mobile', value: '42px'}` (I think this s my favorite :) )
3. Suggesting the consumer trough docs to use constants so they can reuse it everywhere else. 

Please, let me know what do you think? :)

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: Yes, I very much like this naming better. Yeah, lets decide on the API and then I'll apply these changes. :)

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: hey @rbardini, I decided to work on a proposal for you on the same line of thought you suggested, with a minor tweak as I mentioned in the option 2. Please let me know what do you think? :)

File: src/containers/PaddingSelector.tsx
Comment (joanrm20) [reply src/containers/PaddingSelector.tsx]: Indeed. This is a great suggestion. It allowed us to simplify a bit more the `isEnabled` validation.

File: example/.storybook/preview.js
Comment (joanrm20) [inline example/.storybook/preview.js]: hey @rbardini, just opening a space for conversation here on how the API looks like. Please let me know your thoughts. :)

File: src/containers/PaddingSelector.tsx
Comment (joanrm20) [inline src/containers/PaddingSelector.tsx:16]: not necessary anymore as `PaddingWithDefault` is pretty much the same.

File: README.md
Comment (rbardini) [inline README.md:66]: We can remove this `parameters` object now that it's the same example as below.

File: example/.storybook/preview.js
Comment (rbardini) [inline example/.storybook/preview.js]: Let's use camel case names here to match the README examples.

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: Can't we replace `options.values.length > 0` with `isEnabled(options.values)`?

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: I'd prefer to invert the logic and return early when it's disabled:

```tsx
if (!options || options.disable || !isEnabled(options.values)) {
  return [];
}

return options.values.map(...);
```

**Review by rbardini — CHANGES_REQUESTED** [review]
Thanks, @joanrm20, looks much cleaner now 🙂

Added a few minor comments, once they are fixed I'll test locally as well.

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: We could but I think it would not be appropriate, at this moment the `options.values` is not yet `PaddingWithDefault` and actually `isEnabled` should be using the previous and not `Padding`. It didn't complain because it's optional.

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: Sure thing, that can be done.

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: Because we're renaming the `disabled` field` disable`, this will result in a `Uncaught TypeError: Cannot read property 'length' of undefined` for existing users using `disabled`—the execution won't stop at `options.disable`, and `options.values` will be undefined.

I think we need to change the last expression to `!options.values?.length` to maintain backwards compatibility.

**Review by rbardini — CHANGES_REQUESTED** [review]
@joanrm20 I noticed one issue while testing: if you start from a story that disables the addon (e.g. reload from any `Card` story) and navigate to an enabled story, the paddings are not applied, unless you navigate to a second enabled story (or back and forth again).

I couldn't reproduce it on `master`, so it looks like a regression.

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: hey @rbardini, sorry I've been away for awhile. After you upgraded to SB 6 I synced already with master. I've found a way to address the comment you added regarding the regression. I'll test if it works and come back with an update. 😄

File: example/package.json
Comment (rbardini) [inline example/package.json]: Let's use exact versions for dev dependencies.

File: package.json
Comment (rbardini) [inline package.json]: Does `npm run build -- --watch` work?

File: src/types.tsx
Comment (rbardini) [inline src/types.tsx]: Shouldn't this be `string`? Then `PaddingWithDefault` would become `Padding & { default?: boolean }`.

File: src/index.tsx
Comment (rbardini) [inline src/index.tsx:17]: Why do you need to disable? Can't we declare all dependencies?

File: src/helpers.tsx
Comment (rbardini) [inline src/helpers.tsx]: While testing locally, I noticed this is logged twice for every story change. I think we either need to log it only once, or not support arrays at all and just release a major version. What do you think?

**Review by rbardini — COMMENTED** [review]
Looking good, just added some minor suggestions/requests.

Also, it seems Storybook is now padding the preview by default, so even when you clear paddings, there's still some spacing. It's out of scope of this PR though; I'll work on a fix in a separate one.

File: src/helpers.tsx
Comment (joanrm20) [reply src/helpers.tsx]: hey @rbardini, given the last updates on storybook to a major version (and therefore also the addon), considering that the update should be seamless for our consumers (only an API update), I'm therefore leaning more towards just not supporting it anymore.  Will go with the second one.

File: src/index.tsx
Comment (joanrm20) [reply src/index.tsx:17]: Let me take a look on this one, I think this was mostly related to the fact that the name of the component is not named as `use` something. (I think we discussed it briefly somewhere). 

Let me double check a bit and see if I can do anything about it, but i don't think it's related to the deps. 

Edit: correction it's just because of the naming convention. It was asking for `With` while it's named `with`. Although that's the convention for the decorators. At least the a addon backgrounds it's like this.

File: src/types.tsx
Comment (joanrm20) [reply src/types.tsx]: yeah, you're right I missed. Fixed now.

File: package.json
Comment (joanrm20) [reply package.json]: yeah I tested, I think it's useful for sure to have it. I can update to the one you mentioned. Just let me know if you prefer having it?
If not later on we  we can add some documentation to suggest any future maintainer how to watch the build.

File: package.json
Comment (joanrm20) [inline package.json]: this one is fixing a high vulnerability @rbardini

Comment (rbardini) [reply package.json]: Cool, just remember to use exact versions here as well 🙂

File: package.json
Comment (rbardini) [reply package.json]: I think we can keep it, but compose the commands instead 👍

File: src/index.tsx
Comment (joanrm20) [reply src/index.tsx:17]: hey @rbardini, I think this one is the last item, basically this complains as mentioned above, because of the name pattern. Since linting rule it's treating it as a regular component it looks for the Capital case name. What do you think if instead using the `withPaddings` naming, we rename it to `usePaddings`? This way it's compliant with the rule and at the end it's a hook anyway, so it would make sense as well.

Comment (rbardini) [reply src/index.tsx:17]: I checked the backgrounds addon and it follows the same naming convention:

https://github.com/storybookjs/storybook/blob/c8abe86cde37e70d98ef2f5ab3051dbee9df0bf8/addons/backgrounds/src/decorators/withBackground.ts#L6

so let's keep the disable comment for now 🙂

**Comment by rbardini** [comment]
@joanrm20 I just merged #20 to run the main workflow runs on pull requests, and #21 to highlight the selected padding option. Could you please sync your branch to integrate these changes?

**Comment by joanrm20** [comment]
hey @rbardini, done. I have merged and test your latest changes. Please check.

**Review by rbardini — APPROVED** [review]
LGTM! Will merge it soon 👍

File: src/preset/defaultParameters.tsx
Comment (rbardini) [inline src/preset/defaultParameters.tsx]: Let's remove this default, I don't want to be too opinionated 🙂

File: example/stories/2-Card.stories.js
Comment (rbardini) [inline example/stories/2-Card.stories.js]: You can remove this, I'll follow up with a PR refactoring example stories with different use cases.

**Review by rbardini — CHANGES_REQUESTED** [review]
Sorry, found two lines to be removed before we merge this, otherwise looks good 🙂

File: package.json
Comment (rbardini) [inline package.json]: Is `react-dom` needed? I wasn't a dependency before.

Comment (joanrm20) [reply package.json]: hey @rbardini, yeah sort of. I think this comes as part of https://github.com/storybookjs/storybook/pull/13195 SB 6.1 and above. But checking the add-on backgrounds they use it as a peer dependency https://github.com/storybookjs/storybook/blob/next/addons/backgrounds/package.json. we can do the same.

Comment (rbardini) [reply package.json]: I think that would be better, for both React dependencies.

The range they use (`^16.8.0 || ^17.0.0`) is also more inclusive, we can copy that as well 🙂

**Comment by joanrm20** [comment]
hey @rbardini, I've updated it. Please take a look.

File: package.json
Comment (rbardini) [inline package.json]: This version should be fixed.

Comment (joanrm20) [reply package.json]: hey @rbardini done. thanks.

**Review by rbardini — APPROVED** [review]

**Pull request merged**
